posix/device.c

```diff
@@ -0,0 +1,275 @@
+static bool _installed_atexit_handler;
+
+static void _atexit_handler(void)
+{
+    oe_free(_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    if (!_installed_atexit_handler)
+    {
+        oe_atexit(_atexit_handler);
+        _installed_atexit_handler = true;
+    }
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused porition. */
+        memset(p + _table_size, 0, (n - _table_size) * sizeof(entry_t));
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** Public functions:
+**
+**==============================================================================
+*/
+
+static int _clear_devid(uint64_t devid)
+{
+    int ret = -1;
+
+    if (devid >= _table_size || _table[devid] == NULL)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    _table[devid] = NULL;
+
+    ret = 0;
+
+done:
+
+    return ret;
+}
+
+int oe_clear_devid(uint64_t devid)
+{
+    int ret;
+
+    oe_spin_lock(&_lock);
+    ret = _clear_devid(devid);
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_set_device(uint64_t devid, oe_device_t* device)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    if (_resize_table(devid + 1) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    if (_table[devid] != NULL)
+        OE_RAISE_ERRNO(OE_EADDRINUSE);
+
+    _table[devid] = device;
+
+    ret = 0;
+
+done:
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+static oe_device_t* _get_device(uint64_t devid, oe_device_type_t type)
+{
+    oe_device_t* ret = NULL;
+    oe_device_t* device;
+
+    if (devid >= _table_size)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    device = _table[devid];
+
+    if (device && type != OE_DEVICE_TYPE_ANY && device->type != type)
+        goto done;
+
+    ret = device;
+
+done:
+
+    return ret;
+}
+
+oe_device_t* oe_get_device(uint64_t devid, oe_device_type_t type)
+{
+    oe_device_t* ret;
+
+    oe_spin_lock(&_lock);
+    ret = _get_device(devid, type);
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+oe_device_t* oe_find_device(const char* name, oe_device_type_t type)
+{
+    oe_device_t* ret = NULL;
+    oe_device_t* device = NULL;
+    size_t i;
+
+    oe_spin_lock(&_lock);
+
+    if (!name)
+        goto done;
+
+    for (i = 0; i < _table_size; i++)
+    {
+        oe_device_t* p = _table[i];
+
+        if (p && oe_strcmp(p->name, name) == 0)
+        {
+            device = p;
+            break;
+        }
+    }
+
+    if (device && type != OE_DEVICE_TYPE_NONE && device->type != type)
```

Comment:
Consider `OE_DEVICE_TYPE_ANY` for consistency

Reply:
Fixed.